mavlink-core/src/types.rs

```diff
@@ -0,0 +1,147 @@
+use core::ops::{Deref, DerefMut};
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+#[cfg(feature = "arbitrary")]
+use arbitrary::{Arbitrary, Unstructured};
+
+/// Abstraction around a byte array that represents a string.
+///
+/// MAVLink encodes strings as C char arrays and the handling is field dependent.
+/// This abstration allows to choose if one wants to handle the field as
+/// a raw byte array or if one wants the convenience of a str that stops at the first null byte.
+///
+/// # Example
+/// ```
+/// use mavlink_core::types::CharArray;
+///
+/// let data = [0x48, 0x45, 0x4c, 0x4c, 0x4f, 0x00, 0x57, 0x4f, 0x52, 0x4c, 0x44, 0x00, 0x00, 0x00];
+/// let ca = CharArray::new(data);
+/// assert_eq!(ca.to_str(), "HELLO");
+/// ```
+#[derive(Debug, PartialEq, Clone, Copy)]
+#[cfg_attr(feature = "serde", derive(Serialize))]
+#[cfg_attr(feature = "serde", serde(transparent))]
+pub struct CharArray<const N: usize> {
+    #[cfg_attr(feature = "serde", serde(with = "serde_arrays"))]
+    data: [u8; N],
+
+    #[cfg_attr(feature = "serde", serde(skip))]
+    str_len: usize,
+}
+
+impl<const N: usize> CharArray<N> {
+    pub const fn new(data: [u8; N]) -> Self {
+        // Note: The generated code uses this in const contexts, so this is a const fn
+        // and so we can't use iterators or other fancy stuff unfortunately.
+        let mut first_null = N;
+        let mut i = 0;
+        loop {
+            if i >= N {
+                break;
+            }
+            if data[i] == 0 {
+                first_null = i;
+                break;
+            }
+            i += 1;
+        }
+        Self {
+            data,
+            str_len: first_null,
+        }
+    }
+
+    /// Get the string representation of the char array.
+    /// Returns the string stopping at the first null byte and if the string is not valid utf8
+    /// the returned string will be empty.
+    pub fn to_str(&self) -> &str {
+        core::str::from_utf8(&self.data[..self.str_len]).unwrap_or("")
```

Comment:
Why is that we return empty string if the value is invalid? I think the caller should decide what to do on invalid UTF8 values. We can simply return the inner `Result`.